File: Wayfork/WayforkCore/Sources/WayforkCore/Rules/Punycode.swift
import Foundation

/// RFC 3492 Punycode encoder, used to turn IDN labels into their `xn--` form.
enum Punycode {
    private static let base = 36
    private static let tMin = 1
    private static let tMax = 26
    private static let skew = 38
    private static let damp = 700
    private static let initialBias = 72
    private static let initialN = 128

    /// Encodes one label. Returns nil on overflow (labels far beyond DNS limits).
    static func encode(_ label: String) -> String? {
        let input = label.unicodeScalars.map { Int($0.value) }
        var output = input.filter { $0 < 0x80 }.map { Character(Unicode.Scalar(UInt8($0))) }
        let basicCount = output.count
        var handled = basicCount
        if basicCount > 0 {
            output.append("-")
        }

        var n = initialN
        var delta = 0
        var bias = initialBias

        while handled < input.count {
            guard let m = input.filter({ $0 >= n }).min() else { return nil }
            let (product, overflow) = (m - n).multipliedReportingOverflow(by: handled + 1)
            guard !overflow else { return nil }
            delta += product
            n = m
            for c in input {
                if c < n {
                    delta += 1
                }
                guard c == n else { continue }
                var q = delta
                var k = base
                while true {
                    let t = k <= bias ? tMin : (k >= bias + tMax ? tMax : k - bias)
                    if q < t { break }
                    output.append(digit(t + (q - t) % (base - t)))
                    q = (q - t) / (base - t)
                    k += base
                }
                output.append(digit(q))
                bias = adapt(delta: delta, numPoints: handled + 1, firstTime: handled == basicCount)
                delta = 0
                handled += 1
            }
            delta += 1
            n += 1
        }
        return String(output)
    }

    /// `xn--` form of a label, or the label itself when it is pure ASCII.
    static func toASCII(_ label: String) -> String? {
        if label.unicodeScalars.allSatisfy({ $0.isASCII }) {
            return label
        }
        guard let encoded = encode(label) else { return nil }
        return "xn--" + encoded
    }

    private static func digit(_ d: Int) -> Character {
        d < 26 ? Character(Unicode.Scalar(UInt8(97 + d))) : Character(Unicode.Scalar(UInt8(22 + d)))
    }

    private static func adapt(delta: Int, numPoints: Int, firstTime: Bool) -> Int {
        var delta = firstTime ? delta / damp : delta / 2
        delta += delta / numPoints
        var k = 0
        while delta > ((base - tMin) * tMax) / 2 {
            delta /= base - tMin
            k += base
        }
        return k + (base - tMin + 1) * delta / (delta + skew)
    }
}

File: windows/app/lib/core/rules/punycode.dart
/// RFC 3492 Punycode encoder, used to turn IDN labels into their `xn--` form.
abstract final class Punycode {
  static const _base = 36;
  static const _tMin = 1;
  static const _tMax = 26;
  static const _skew = 38;
  static const _damp = 700;
  static const _initialBias = 72;
  static const _initialN = 128;
  static const _maxInt64 = 0x7fffffffffffffff;

  /// Encodes one label. Returns null on overflow (labels far beyond DNS limits).
  static String? encode(String label) {
    final input = label.runes.toList();
    final output = <int>[...input.where((value) => value < 0x80)];
    final basicCount = output.length;
    var handled = basicCount;
    if (basicCount > 0) output.add(0x2d);

    var n = _initialN;
    var delta = 0;
    var bias = _initialBias;
    while (handled < input.length) {
      int? m;
      for (final value in input) {
        if (value >= n && (m == null || value < m)) m = value;
      }
      if (m == null) return null;
      final difference = m - n;
      if (difference > _maxInt64 ~/ (handled + 1)) return null;
      final product = difference * (handled + 1);
      if (delta > _maxInt64 - product) return null;
      delta += product;
      n = m;
      for (final value in input) {
        if (value < n) {
          if (delta == _maxInt64) return null;
          delta++;
        }
        if (value != n) continue;
        var q = delta;
        var k = _base;
        while (true) {
          final t = k <= bias ? _tMin : (k >= bias + _tMax ? _tMax : k - bias);
          if (q < t) break;
          output.add(_digit(t + (q - t) % (_base - t)));
          q = (q - t) ~/ (_base - t);
          k += _base;
        }
        output.add(_digit(q));
        bias = _adapt(delta, handled + 1, handled == basicCount);
        delta = 0;
        handled++;
      }
      if (delta == _maxInt64 || n == _maxInt64) return null;
      delta++;
      n++;
    }
    return String.fromCharCodes(output);
  }

  /// `xn--` form of a label, or the label itself when it is pure ASCII.
  static String? toASCII(String label) {
    if (label.runes.every((value) => value < 0x80)) return label;
    final encoded = encode(label);
    return encoded == null ? null : 'xn--$encoded';
  }

  static int _digit(int value) => value < 26 ? 97 + value : 22 + value;

  static int _adapt(int value, int numPoints, bool firstTime) {
    var delta = firstTime ? value ~/ _damp : value ~/ 2;
    delta += delta ~/ numPoints;
    var k = 0;
    while (delta > ((_base - _tMin) * _tMax) ~/ 2) {
      delta ~/= _base - _tMin;
      k += _base;
    }
    return k + (_base - _tMin + 1) * delta ~/ (delta + _skew);
  }
}
